File: backend/app/api/leadership.py
"""Leadership's named comparison sets: the seven routes the definition surface is (SPEC §5.1).

§13's tree gives the leadership-facing API this module, and §13's closing rule
keeps it thin: a handler here resolves the session, hands the work to
`app.services.comparison_sets`, and turns what comes back into an HTTP answer.
Every decision — who may change which set, what a set may be, how wide one is —
is in that service, and the payload's shape is `app.schemas.comparison_sets`.

**Seven routes and no others.** The institution's sets, one set, the choices a
set is defined out of, and the three writes; plus the preview, which is how a
definer sees what a set reaches before anything renders a figure over it. E5-09
consumes all seven.

**`/leadership/comparison-sets/options` is declared before
`/leadership/comparison-sets/{set_id}`, and that is load-bearing.** FastAPI
matches in declaration order, so with the keyed route first this path is read as
a set id that is not a uuid and the whole definition form answers a validation
error. The two are written in that order here and
`tests/integration/test_the_named_set_options_offer_only_the_valid_choices.py`
holds them in it.

**Every route carries one of `app.api.deps`' two leadership dependencies rather
than a check of its own.** That is what makes them findable: a sweep asking the
running application which routes carry `require_leadership` gets this module's
whole surface, and a route that resolved a session for itself would be a
leadership route outside every such walk — the shape `app.api.student` and
`app.api.instructor` already hold. The reads carry `require_leadership` and the
three writes carry `csrf_verified_leadership`, which is the same gate plus ADR
0089's double-submit check.

**Scope is asymmetric, and the asymmetry is the decision** (ADR 0173). Every
leadership session reads every set in the institution — §5.1's set-definition
surface is one list, and two leaders looking at different halves of it would
define the same cohort twice. Only the leader whose `person` row defined a set
may edit or delete it, which is what `editable` on the list says and what the
403 below answers. Purview over the supervision graph is E9's.

**Three refusals and three layers, which is why they are three sentences.** A
session that is not a leadership session is refused by the dependency with a 401
and a `Bearer` challenge; a set id nothing defined is a 404; another leader's set
on a write is a 403. The 404 comes before the 403 on a write, deliberately: a set
that is not there is not there for anybody, and answering 403 for an unknown id
would tell a caller that a set exists and belongs to somebody else — the
enumeration `app.api.instructor`'s own refusal pair exists to prevent, in the one
place here where it has a foothold.

**The five write refusals are translations rather than checks.** SPEC §2.2's
lengths, §8's levels and §5.1's exact level match are all held by Postgres
(E5-01, ADR 0164); the service attempts the write and maps the constraint that
fired to one sentence. So a 422 from here carries a sentence out of
`app.copy.leadership_sets` rather than a list of field errors, and a body that
carried field errors would mean the wire model had refused the value before the
database saw it.

**The preview answers two counts and nothing else.** SPEC §4.1 item 7 suppresses
statistics computed over a comparison set below the configured minimums; a count
of member courses and a count of resolved sections say how wide a cohort is and
nothing about what anybody in it answered, which is what makes them safe at
definition time. A name, a section code or a mean here would be a figure reaching
a reader outside E4-07's suppression chokepoint.

**Every answer carries `Cache-Control: no-store`.** A leadership answer sitting
in a shared cache is a set definition, and the list of every cohort in the
institution, served to whoever asks next on the same machine. The student and
instructor read paths set the same header for the same reason.

**No `audit_log` row is written by any of this** (ADR 0174). What is recorded is
on the set's own row — the creator and `created_at` on a create, `updated_at` on
an edit — and `docs/tickets/e5/deferred.md` carries what a delete leaves behind,
which is nothing.
"""

from uuid import UUID

from fastapi import APIRouter, Depends, HTTPException, Request, Response
from sqlalchemy.orm import Session

from app.api.deps import csrf_verified_leadership, require_leadership
from app.config import Settings
from app.copy.leadership_sets import NOT_THE_SETS_DEFINER, SET_UNAVAILABLE
from app.db import get_session
from app.schemas.comparison_sets import (
    SetDetail,
    SetList,
    SetOptions,
    SetPreview,
    SetWrite,
)
from app.services.comparison_sets import (
    NotTheSetsDefinerError,
    SetUnavailableError,
    WriteRefusedError,
    create_set,
    definition_options,
    delete_set,
    edit_set,
    listed_sets,
    preview_of,
    read_set,
)
from app.services.session import SessionClaims

router = APIRouter(tags=["leadership"])

# Where the institution's sets are listed and where one is defined. Written out
# in full rather than carried on a router prefix, so that what a route is
# registered at is what this file says it is — the convention `app.api.student`
# settled and `app.api.instructor` follows.
SETS_PATH = "/leadership/comparison-sets"

# The choices a set is defined out of. **Declared before the keyed paths below**
# — see the module docstring — and written out in full like its siblings.
OPTIONS_PATH = f"{SETS_PATH}/options"

# One set, and what it reaches.
SET_PATH = f"{SETS_PATH}/{{set_id}}"
PREVIEW_PATH = f"{SET_PATH}/preview"

# The statuses the three keyed refusals answer with. 404 for an id nothing
# defined and 403 for another leader's set: the two are different facts and a
# leader may read every set, so saying "this one is not yours" discloses nothing
# the list does not already show.
SET_UNAVAILABLE_STATUS = 404
NOT_THE_SETS_DEFINER_STATUS = 403

CREATED = 201
NO_CONTENT = 204

NO_STORE = "no-store"


@router.get(SETS_PATH, summary="Every comparison set this institution has defined")
def read_sets(
    response: Response,
    claims: SessionClaims = Depends(require_leadership),
    session: Session = Depends(get_session),
) -> SetList:
    """The institution's sets, in name order, each marked editable or not.

    **Institution-wide rather than this session's own** (ADR 0173). The reader
    still comes from the session and from nowhere else: `editable` is decided
    against the `person` row this session was resolved to at the door, so no part
    of this request's own text decides which sets it may change.

    **Synchronous, and FastAPI runs it in a threadpool**, for the reason
    `app.api.instructor` gives: the session is synchronous (ADR 0013) and every
    statement behind this blocks, so an `async` handler would take them on the
    event loop.
    """
    response.headers["Cache-Control"] = NO_STORE
    return SetList(sets=listed_sets(session, person_id=_person_of(claims)))


@router.post(SETS_PATH, status_code=CREATED, summary="Define a comparison set")
def define_set(
    write: SetWrite,
    request: Request,
    response: Response,
    claims: SessionClaims = Depends(csrf_verified_leadership),
    session: Session = Depends(get_session),
) -> SetDetail:
    """Define a set for whoever this session is, or translate what the database refused.

    **The definer comes from the session and never from the body**, and it is
    what every later edit and delete of this set is scoped by.
    """
    settings: Settings = request.app.state.settings
    response.headers["Cache-Control"] = NO_STORE
    try:
        return create_set(session, write=write, person_id=_person_of(claims), settings=settings)
    except NotTheSetsDefinerError:
        raise _not_the_definer() from None
    except WriteRefusedError as refused:
        raise _refused(refused) from None


@router.get(OPTIONS_PATH, summary="The lengths, levels and courses a set may be defined out of")
def read_options(
    response: Response,
    claims: SessionClaims = Depends(require_leadership),
    session: Session = Depends(get_session),
) -> SetOptions:
    """The closed choices the definition form offers. Declared before `{set_id}`.

    It takes no parameter, so there is nothing in the request to refuse and no
    refusal here beyond the role gate. The courses it lists are every course this
    institution runs, because the form is where a new cohort is built.
    """
    response.headers["Cache-Control"] = NO_STORE
    return definition_options(session)


@router.get(SET_PATH, summary="One comparison set, whoever defined it")
def read_one_set(
    set_id: UUID,
    response: Response,
    claims: SessionClaims = Depends(require_leadership),
    session: Session = Depends(get_session),
) -> SetDetail:
    """One set and its membership.

    **A set id that is not a uuid is refused before this runs**, by FastAPI's own
    parsing of the path parameter — which is an answer about the shape of a
    request rather than about a set. ADR 0016 makes every key a uuid, so a
    malformed value is not a set id at all and never reaches the lookup.
    """
    response.headers["Cache-Control"] = NO_STORE
    try:
        return read_set(session, set_id=set_id, person_id=_person_of(claims))
    except SetUnavailableError:
        raise _unavailable() from None


@router.put(SET_PATH, summary="Replace the definition of a set I defined")
def replace_set(
    set_id: UUID,
    write: SetWrite,
    request: Request,
    response: Response,
    claims: SessionClaims = Depends(csrf_verified_leadership),
    session: Session = Depends(get_session),
) -> SetDetail:
    """Replace a set's whole definition, for the leader who defined it.

    The three refusals in order: an id nothing defined is a 404, another leader's
    set is a 403, and a definition the database will not store is one of the five
    translated sentences.
    """
    settings: Settings = request.app.state.settings
    response.headers["Cache-Control"] = NO_STORE
    try:
        return edit_set(
            session,
            set_id=set_id,
            write=write,
            person_id=_person_of(claims),
            settings=settings,
        )
    except SetUnavailableError:
        raise _unavailable() from None
    except NotTheSetsDefinerError:
        raise _not_the_definer() from None
    except WriteRefusedError as refused:
        raise _refused(refused) from None


@router.delete(SET_PATH, status_code=NO_CONTENT, summary="Delete a set I defined")
def remove_set(
    set_id: UUID,
    response: Response,
    claims: SessionClaims = Depends(csrf_verified_leadership),
    session: Session = Depends(get_session),
) -> None:
    """Delete a set, taking its membership rows and nothing else (ADR 0164).

    204 and no body, because there is nothing left to describe. A course named by
    the set is untouched and stays undeletable while any set names it, which is
    the opposite direction of the same rule.
    """
    response.headers["Cache-Control"] = NO_STORE
    try:
        delete_set(session, set_id=set_id, person_id=_person_of(claims))
    except SetUnavailableError:
        raise _unavailable() from None
    except NotTheSetsDefinerError:
        raise _not_the_definer() from None


@router.get(PREVIEW_PATH, summary="How many courses and sections a set reaches")
def read_preview(
    set_id: UUID,
    response: Response,
    claims: SessionClaims = Depends(require_leadership),
    session: Session = Depends(get_session),
) -> SetPreview:
    """Two counts and nothing else — see the module docstring on what that costs and why.

    Answered for every set, not only for this session's own: a count is what
    makes an institution-wide list usable, and it says nothing about anybody.
    """
    response.headers["Cache-Control"] = NO_STORE
    try:
        return preview_of(session, set_id=set_id)
    except SetUnavailableError:
        raise _unavailable() from None


def _person_of(claims: SessionClaims) -> UUID | None:
    """The `person` row this session was resolved to, or `None` where it names none.

    A claim in a JWT is JSON, so `person_id` is a string here and a `uuid.UUID`
    everywhere below (ADR 0016). A value that is not one is a token this
    deployment did not issue in the shape it issues them, and it resolves to
    nobody rather than to a 500 from inside the parse — which the service answers
    as a session that may read every set and write none.

    The same reader `app.api.instructor` holds, and it is deliberately a second
    copy of four lines rather than a shared helper: promoting it is a change to a
    module this ticket was told not to touch, and the pull request proposes it
    rather than making it.
    """
    if claims.person_id is None:
        return None
    try:
        return UUID(claims.person_id)
    except ValueError:
        return None


def _unavailable() -> HTTPException:
    """The 404 for a set id nothing defined, carrying one sentence and nothing it was handed."""
    return HTTPException(status_code=SET_UNAVAILABLE_STATUS, detail=SET_UNAVAILABLE)


def _not_the_definer() -> HTTPException:
    """The 403 for another leader's set on a write."""
    return HTTPException(status_code=NOT_THE_SETS_DEFINER_STATUS, detail=NOT_THE_SETS_DEFINER)


def _refused(refused: WriteRefusedError) -> HTTPException:
    """One translated constraint refusal, with the status that rule is answered under."""
    return HTTPException(status_code=refused.status_code, detail=refused.detail)

File: frontend/src/api/leadership.ts
/**
 * The seven calls the comparison-set management screen makes — SPEC §13's
 * `frontend/src/api/`, ticket E5-09.
 *
 * `api/leadership.py` (E5-06) serves SPEC §5.1's named sets: a list, a create, a
 * read, an edit, a delete, the closed choice lists the form is built from, and
 * the preview that says what a set reaches. This module is the client half, and
 * it is the third in this directory rather than a growth of either of the other
 * two: `instructor.ts` reads one instructor's own report and `student.ts` is one
 * student's survey, and neither has anything to say about a leadership surface.
 *
 * **Hand-written, not generated, and not wrapped in a query cache** — ADR 0117,
 * unchanged, and this screen is the case it was argued for: seven calls with no
 * shared cache key between them.
 *
 * **The session rides as a Bearer header** (ADR 0089, `../lib/session.ts`), and
 * **every write echoes the double-submit cookie** through `csrfHeader()`. This
 * is the first client in this directory with writes other than the student's
 * submission, which is why that reader moved out of `student.ts` and up into
 * `lib/session.ts` rather than being copied here (`docs/MISTAKES.md` entry 13).
 *
 * **What checks that echo is `csrf_verified_leadership`** (`api/deps.py`,
 * E5-06), which the three writing routes below declare and the four reading
 * routes do not.
 *
 * **Every field below is the wire's spelling**, snake case included, because
 * these types describe E5-06's Pydantic schemas rather than a shape of this
 * screen's choosing.
 *
 * **A refusal is shown, never re-derived.** SPEC §5.1 puts the rules about which
 * length and level may be combined, who may edit a set and what a duplicate name
 * means on the server; `api/leadership.py` answers each with one sentence under
 * `detail`, and the screen renders the sentence it was sent. The form makes the
 * invalid combinations unreachable for a person using it — that is §5.1's "makes
 * invalid combinations impossible rather than erroring on them" — and the
 * refusals stay reachable by the races a form cannot prevent: a course that left
 * the catalogue while the form was open, a name somebody else took first.
 */

import { authorizationHeader, csrfHeader } from '../lib/session';

/** `app.api.leadership.COMPARISON_SETS_PATH` — the list, and the create. */
export const COMPARISON_SETS_PATH = '/leadership/comparison-sets';

/** The closed choice lists the form is built from. */
export const COMPARISON_SET_OPTIONS_PATH = `${COMPARISON_SETS_PATH}/options`;

/** One set: the read, the edit and the delete all address it. */
export function comparisonSetPath(setId: string): string {
  return `${COMPARISON_SETS_PATH}/${encodeURIComponent(setId)}`;
}

/** What one set reaches: its member count and its resolved section count. */
export function comparisonSetPreviewPath(setId: string): string {
  return `${comparisonSetPath(setId)}/preview`;
}

/** One course the form may offer, in the level band that decides where it appears. */
export interface ComparisonSetCourseView {
  readonly id: string;
  /** The governed label the server composes; nothing here assembles one. */
  readonly label: string;
  /** One of SPEC §8's five bands, as the server spells it. */
  readonly level: string;
}

/**
 * The closed sets the form offers, and nothing else.
 *
 * **This is the only source of the lengths and the levels.** The lengths are
 * the distinct lengths the institution's sections run, read from its data, and
 * §8 owns the five level bands; both live on the server's side of the wire. A
 * list written into a component would be a second copy that is right until the
 * first section that runs a new length —
 * the two-currencies defect this ticket's trap section names — so the form
 * renders what this answer carries and has no opinion about what it should have
 * carried.
 */
export interface ComparisonSetOptionsView {
  readonly lengths: readonly number[];
  readonly levels: readonly string[];
  readonly courses: readonly ComparisonSetCourseView[];
}

/** One set as the list carries it. */
export interface ComparisonSetSummaryView {
  readonly id: string;
  readonly name: string;
  readonly length_weeks: number;
  readonly level: string;
  readonly member_count: number;
  /**
   * Whether this reader may edit and delete this set.
   *
   * The server's answer to a question about who defined the set and what this
   * session's person may do with it, and it is not re-derived here: a screen
   * deciding for itself which controls to draw would be a second authority on a
   * question SPEC §2.1 gives to one. The controls follow the flag; the route
   * still refuses a write the flag should have withheld.
   */
  readonly editable: boolean;
}

/** One set with its membership, as the read and both writes answer. */
export interface ComparisonSetDetailView extends ComparisonSetSummaryView {
  readonly member_course_ids: readonly string[];
  readonly created_at: string;
  readonly updated_at: string;
}

/** What a create or an edit sends. */
export interface ComparisonSetWrite {
  readonly name: string;
  readonly length_weeks: number;
  readonly level: string;
  readonly member_course_ids: readonly string[];
}

/**
 * What a set reaches.
 *
 * `section_count` is optional because E5-06's scope note allows the preview to
 * answer the member count without it. An absent count is rendered as an absence;
 * a zero printed in its place would state something the server did not say.
 */
export interface ComparisonSetPreviewView {
  readonly member_count: number;
  readonly section_count?: number | null;
}

/** What a read of the set list answered. */
export type ComparisonSetsRead =
  | { readonly kind: 'sets'; readonly sets: readonly ComparisonSetSummaryView[] }
  | { readonly kind: 'session-ended' }
  | { readonly kind: 'unavailable'; readonly detail: string | null };

/** What a read of the choice lists answered. */
export type ComparisonSetOptionsRead =
  | { readonly kind: 'options'; readonly options: ComparisonSetOptionsView }
  | { readonly kind: 'session-ended' }
  | { readonly kind: 'unavailable'; readonly detail: string | null };

/** What a read of one set answered. */
export type ComparisonSetRead =
  | { readonly kind: 'set'; readonly set: ComparisonSetDetailView }
  | { readonly kind: 'not-found'; readonly detail: string | null }
  | { readonly kind: 'session-ended' }
  | { readonly kind: 'unavailable'; readonly detail: string | null };

/** What a read of one set's preview answered. */
export type ComparisonSetPreviewRead =
  | { readonly kind: 'preview'; readonly preview: ComparisonSetPreviewView }
  | { readonly kind: 'not-found'; readonly detail: string | null }
  | { readonly kind: 'session-ended' }
  | { readonly kind: 'unavailable'; readonly detail: string | null };

/**
 * What a create or an edit answered.
 *
 * **The four refusals are one outcome on purpose.** A duplicate name, a set this
 * person may not edit, a set that is not there and a member the server will not
 * accept are 409, 403, 404 and 422, and each carries its own sentence. The
 * screen shows the sentence; it does not tell the four apart, because telling
 * them apart would mean holding a second copy of the rules that decide them.
 */
export type ComparisonSetWriteOutcome =
  | { readonly kind: 'saved'; readonly set: ComparisonSetDetailView }
  | { readonly kind: 'refused'; readonly detail: string | null }
  | { readonly kind: 'session-ended' }
  | { readonly kind: 'unavailable'; readonly detail: string | null };

/** What a delete answered. The same shape, with nothing to carry back. */
export type ComparisonSetDeleteOutcome =
  | { readonly kind: 'deleted' }
  | { readonly kind: 'refused'; readonly detail: string | null }
  | { readonly kind: 'session-ended' }
  | { readonly kind: 'unavailable'; readonly detail: string | null };

/** No leadership session on the request (`require_leadership`). */
const UNAUTHORIZED_STATUS = 401;

/** The set is not there, or is not one this reader may read. */
const NOT_FOUND_STATUS = 404;

/** The headers a read carries. */
function readHeaders(): Record<string, string> {
  return { Accept: 'application/json', ...authorizationHeader() };
}

/**
 * The headers a write carries: the session, the body's type and the
 * double-submit echo when the cookie is readable.
 */
function writeHeaders(): Record<string, string> {
  return {
    Accept: 'application/json',
    'Content-Type': 'application/json',
    ...authorizationHeader(),
    ...csrfHeader(),
  };
}

/** A response's JSON body, or `null` when it did not carry one. */
async function jsonBody(response: Response): Promise<unknown> {
  try {
    return (await response.json()) as unknown;
  } catch {
    return null;
  }
}

/**
 * `detail` out of an error body, when it is a sentence.
 *
 * FastAPI answers a refusal with `{"detail": …}`, and every refusal these routes
 * serve carries a **string** there. FastAPI's own 422 for a body it could not
 * parse carries a list of validation objects instead, which is not a sentence
 * anybody wrote for a reader, so it is answered `null` here and the screen uses
 * its own words.
 */
function refusalSentence(body: unknown): string | null {
  if (typeof body !== 'object' || body === null) return null;
  const detail = (body as Record<string, unknown>).detail;
  return typeof detail === 'string' ? detail : null;
}

/** The sets this session's person may read, by name. */
export async function readComparisonSets(): Promise<ComparisonSetsRead> {
  let response: Response;
  try {
    response = await fetch(COMPARISON_SETS_PATH, { headers: readHeaders() });
  } catch {
    return { kind: 'unavailable', detail: null };
  }

  if (response.status === UNAUTHORIZED_STATUS) return { kind: 'session-ended' };
  const body = await jsonBody(response);
  if (!response.ok) return { kind: 'unavailable', detail: refusalSentence(body) };

  // The bound on the cast: `sets` is the member every render walks, and a body
  // without it would fail somewhere deeper with nothing to say.
  if (typeof body !== 'object' || body === null) return { kind: 'unavailable', detail: null };
  if (!Array.isArray((body as Record<string, unknown>).sets)) {
    return { kind: 'unavailable', detail: null };
  }
  return { kind: 'sets', sets: (body as { sets: readonly ComparisonSetSummaryView[] }).sets };
}

/** The closed choice lists the form offers. */
export async function readComparisonSetOptions(): Promise<ComparisonSetOptionsRead> {
  let response: Response;
  try {
    response = await fetch(COMPARISON_SET_OPTIONS_PATH, { headers: readHeaders() });
  } catch {
    return { kind: 'unavailable', detail: null };
  }

  if (response.status === UNAUTHORIZED_STATUS) return { kind: 'session-ended' };
  const body = await jsonBody(response);
  if (!response.ok) return { kind: 'unavailable', detail: refusalSentence(body) };

  // All three members are walked by the form, and a form built from a partial
  // answer would offer a choice list the server never sent.
  if (typeof body !== 'object' || body === null) return { kind: 'unavailable', detail: null };
  const options = body as Record<string, unknown>;
  if (
    !Array.isArray(options.lengths) ||
    !Array.isArray(options.levels) ||
    !Array.isArray(options.courses)
  ) {
    return { kind: 'unavailable', detail: null };
  }
  return { kind: 'options', options: body as ComparisonSetOptionsView };
}

/** One set, with the courses in it. */
export async function readComparisonSet(setId: string): Promise<ComparisonSetRead> {
  let response: Response;
  try {
    response = await fetch(comparisonSetPath(setId), { headers: readHeaders() });
  } catch {
    return { kind: 'unavailable', detail: null };
  }

  if (response.status === UNAUTHORIZED_STATUS) return { kind: 'session-ended' };
  const body = await jsonBody(response);
  if (response.status === NOT_FOUND_STATUS) {
    return { kind: 'not-found', detail: refusalSentence(body) };
  }
  if (!response.ok) return { kind: 'unavailable', detail: refusalSentence(body) };

  if (typeof body !== 'object' || body === null) return { kind: 'unavailable', detail: null };
  if (!Array.isArray((body as Record<string, unknown>).member_course_ids)) {
    return { kind: 'unavailable', detail: null };
  }
  return { kind: 'set', set: body as ComparisonSetDetailView };
}

/** What one set reaches, for the list's preview line. */
export async function readComparisonSetPreview(setId: string): Promise<ComparisonSetPreviewRead> {
  let response: Response;
  try {
    response = await fetch(comparisonSetPreviewPath(setId), { headers: readHeaders() });
  } catch {
    return { kind: 'unavailable', detail: null };
  }

  if (response.status === UNAUTHORIZED_STATUS) return { kind: 'session-ended' };
  const body = await jsonBody(response);
  if (response.status === NOT_FOUND_STATUS) {
    return { kind: 'not-found', detail: refusalSentence(body) };
  }
  if (!response.ok) return { kind: 'unavailable', detail: refusalSentence(body) };

  // `member_count` is the one member the preview always carries; the section
  // count is allowed to be absent and the line says so in words when it is.
  if (typeof body !== 'object' || body === null) return { kind: 'unavailable', detail: null };
  if (typeof (body as Record<string, unknown>).member_count !== 'number') {
    return { kind: 'unavailable', detail: null };
  }
  return { kind: 'preview', preview: body as ComparisonSetPreviewView };
}

/** Define a set. */
export async function createComparisonSet(
  write: ComparisonSetWrite,
): Promise<ComparisonSetWriteOutcome> {
  return await sendComparisonSet(COMPARISON_SETS_PATH, 'POST', write);
}

/** Edit a set that already exists. */
export async function updateComparisonSet(
  setId: string,
  write: ComparisonSetWrite,
): Promise<ComparisonSetWriteOutcome> {
  return await sendComparisonSet(comparisonSetPath(setId), 'PUT', write);
}

/**
 * The body of both writes, which differ only in address and method.
 *
 * One function rather than two near-identical ones: what a create and an edit
 * each do with a refusal, with a 401 and with the answer is the same decision,
 * and two copies of it are two places for that decision to drift.
 */
async function sendComparisonSet(
  path: string,
  method: 'POST' | 'PUT',
  write: ComparisonSetWrite,
): Promise<ComparisonSetWriteOutcome> {
  let response: Response;
  try {
    response = await fetch(path, {
      method,
      headers: writeHeaders(),
      body: JSON.stringify(write),
    });
  } catch {
    return { kind: 'unavailable', detail: null };
  }

  if (response.status === UNAUTHORIZED_STATUS) return { kind: 'session-ended' };
  const body = await jsonBody(response);
  if (!response.ok) return { kind: 'refused', detail: refusalSentence(body) };

  if (typeof body !== 'object' || body === null) return { kind: 'unavailable', detail: null };
  if (!Array.isArray((body as Record<string, unknown>).member_course_ids)) {
    return { kind: 'unavailable', detail: null };
  }
  return { kind: 'saved', set: body as ComparisonSetDetailView };
}

/** Delete a set. */
export async function deleteComparisonSet(setId: string): Promise<ComparisonSetDeleteOutcome> {
  let response: Response;
  try {
    response = await fetch(comparisonSetPath(setId), {
      method: 'DELETE',
      headers: writeHeaders(),
    });
  } catch {
    return { kind: 'unavailable', detail: null };
  }

  if (response.status === UNAUTHORIZED_STATUS) return { kind: 'session-ended' };
  if (response.ok) return { kind: 'deleted' };
  return { kind: 'refused', detail: refusalSentence(await jsonBody(response)) };
}
